### PR TITLE
8301564: Non-C-heap allocated ResourceHashtable keys and values must have trivial destructor

### PR DESCRIPTION
**Pull request by iklam**
To ensure we don't have memory leaks or other more serious memory management bugs, I added static asserts to check that the `K` and `V` types for `ResourceHashtableBase` must have trivial destructors if the table is not `C_HEAP` allocated. Thanks to @JornVernee for the original assert code.

The asserts actually found a problem with `ClassLoaderStatsClosure::StatsTable`. The space used by the `oop` in the freed hashtable may be trashed when `-XX:+CheckUnhandledOops` is enabled, so live objects that reuse the same space may be corrupted. (I tried but couldn't get it to fail).

I also had to change `CodeBuffer::SharedTrampolineRequests` because it's `V` type is `LinkListImpl<int>`, which has a non-trivial destructor. (Note: in this case, the destructor doesn't do anything; we can probably make it go away with C++-20. See https://stackoverflow.com/questions/40094871/sfinae-away-a-destructor)

Testing with tier1~4 + builds-tier5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301564](https://bugs.openjdk.org/browse/JDK-8301564): Non-C-heap allocated ResourceHashtable keys and values must have trivial destructor

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [e3ccac72](https://git.openjdk.org/jdk/pull/12355/files/e3ccac7233a8a140b0489f6896f7893aa6f660ac)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [e3ccac72](https://git.openjdk.org/jdk/pull/12355/files/e3ccac7233a8a140b0489f6896f7893aa6f660ac)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12355/head:pull/12355` \
`$ git checkout pull/12355`

Update a local copy of the PR: \
`$ git checkout pull/12355` \
`$ git pull https://git.openjdk.org/jdk pull/12355/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12355`

View PR using the GUI difftool: \
`$ git pr show -t 12355`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12355.diff">https://git.openjdk.org/jdk/pull/12355.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderStats.hpp [hunk 1]
```diff
@@ -112,7 +112,7 @@ class ClassLoaderStatsClosure : public CLDClosure {
   }
 
   typedef ResourceHashtable<oop, ClassLoaderStats,
-                            256, AnyObj::RESOURCE_AREA, mtInternal,
+                            256, AnyObj::C_HEAP, mtInternal,
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderStats.hpp:115]: There's an mtStatistics NMT category, can you use that for this? and the instance below?

**Review by coleenp — APPROVED** [review]
Looks good. Glad there weren't more hashtables breaking this assumption.

**Review by JornVernee — APPROVED** [review]
This looks good, thanks.

I'll say this for posterity: we discussed changing `ResourceHashtable` to always call the destructor of the key and value (sketch [here](https://github.com/openjdk/jdk/compare/master...JornVernee:jdk:Always_Dtor)). That would allow using `K` and `V` with non-trivial destructors as well. But, this discussion didn't reach a conclusion yet (in particular, there were some doubts expressed over whether a container should be calling destructors at all).

In the mean time, this patch improves upon the status quo.

File: src/hotspot/share/classfile/classLoaderStats.hpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/classfile/classLoaderStats.hpp]: Fixed.

**Comment by iklam** [comment]
Thanks @JornVernee and @coleenp for the review.
/integrate

**Pull request closed without merging**